File: backend/app/schemas/risk.py
from __future__ import annotations

from pydantic import BaseModel, Field


class OrderCheckRequest(BaseModel):
    symbol: str = Field(min_length=6, max_length=6, pattern=r"^\d{6}$")
    side: str = Field(pattern=r"^(buy|sell)$")
    price: float = Field(gt=0)
    quantity: int = Field(gt=0)


class OrderCheckResponse(BaseModel):
    allowed: bool
    max_qty: int
    reasons: list[str]
    warnings: list[str]
    state: str


class PositionParamsModel(BaseModel):
    single_stock_max_pct: float
    total_position_max_pct: float
    strategy_weights: dict[str, float]
    stop_loss_pct: float
    add_position_limit: str
    high_position_stock_limit: str
    drawdown_protection_pct: float


class RiskStatePayload(BaseModel):
    state: str
    reasons: list[str]
    params: PositionParamsModel
    updated_at: str | None = None

File: apps/web/lib/api/risk.ts
/**
 * 风控：状态与下单预检
 *
 * `lib/api.ts` 的内部切片（IMP-005，2026-09-15 从 2751 行单文件按业务域拆出）。
 * **只搬位置、不重写**：声明正文与拆分前逐字符相同；对外仍由 `lib/api.ts` 统一转发，
 * 因此引用方（`@/lib/api`）**零改动**。
 */

import { getJson, sendJson } from "./internal";

export interface RiskState {
  state: string;
  reasons: string[];
  params: {
    single_stock_max_pct: number;
    total_position_max_pct: number;
    strategy_weights: Record<string, number>;
    stop_loss_pct: number;
    add_position_limit: string;
    high_position_stock_limit: string;
    drawdown_protection_pct: number;
  };
  updated_at: string | null;
}

export interface OrderCheckResult {
  allowed: boolean;
  max_qty: number;
  reasons: string[];
  warnings: string[];
  state: string;
}

export interface OrderCheckRequest {
  symbol: string;
  side: "buy" | "sell";
  price: number;
  quantity: number;
}

export async function getRiskState(): Promise<RiskState> {
  return (await getJson<RiskState>("/api/risk/state")).data;
}

export async function checkOrderRisk(req: OrderCheckRequest): Promise<OrderCheckResult> {
  return (await sendJson<OrderCheckResult>("/api/risk/check-order", "POST", req)).data;
}
